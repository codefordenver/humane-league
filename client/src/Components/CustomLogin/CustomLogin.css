.CustomLogin {
  display: grid;
  width: 600px;
  margin: 0px auto;
  grid-template-columns: 80px repeat(4, 1fr);
  grid-gap: 10px;
  background: #ffffffe3;
  padding: 0px 30px 20px 15px;
  border-radius: 2px; }
  .CustomLogin h1 {
    grid-column: span 5;
    text-align: center; }
  .CustomLogin label {
    display: flex;
    align-items: center; }
  .CustomLogin input {
    height: 50px;
    font-size: 24px;
    padding-left: 10px;
    outline: none;
    grid-column: span 4; }
  .CustomLogin span {
    min-height: 20px;
    min-width: 100%;
    color: red;
    font-size: 18px;
    grid-column: span 5;
    text-align: center; }
  .CustomLogin button {
    grid-column: span 2;
    height: 70px;
    display: flex;
    width: 220px;
    align-items: center;
    justify-content: center;
    border-radius: 2px;
    border: none;
    background: #1CB841;
    color: white;
    cursor: pointer;
    margin: 0px auto;
    transition: 0.3s;
    font-size: 20px;
    padding: 10px; }
    .CustomLogin button:nth-of-type(2) {
      grid-column: 4 / span 2;
      justify-self: end; }
    .CustomLogin button:hover {
      color: white;
      background-color: #323232; }
    .CustomLogin button:disabled {
      background: #989696;
      color: black; }
  .CustomLogin .confirmWrapper {
    position: relative;
    grid-column: span 4; }
    .CustomLogin .confirmWrapper input {
      width: 100%; }
    .CustomLogin .confirmWrapper span {
      position: absolute;
      display: flex;
      justify-content: center;
      align-items: center;
      z-index: 10;
      color: black;
      right: 0px;
      top: 0px;
      height: 100%;
      min-width: 50px; }
  .CustomLogin a {
    grid-column: 5 / -1;
    text-align: right;
    font-size: 14px; }
