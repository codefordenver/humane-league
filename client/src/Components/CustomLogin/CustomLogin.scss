.CustomLogin {
  display: grid;
  width: 600px;
  margin: 0px auto;
  grid-template-columns: 80px 1fr;
  grid-gap: 10px;

  h1 {
    grid-column: span 2;
    text-align: center;
  }

  label {
    display: flex;
    align-items: center;
  }

  input {
    height: 50px;
    font-size: 24px;
    padding-left: 10px;
    outline: none;
  }

  span {
    min-height: 15px;
    min-width: 100%;
    color: red;
    font-size: 12px;
    grid-column: span 2;
    text-align: center;
  }

  button,
  a {
    grid-column: span 2;
    height: 70px;
    display: flex;
    width: 250px;
    align-items: center;
    justify-content: center;
    border-radius: 2px;
    border: none;
    background: #1CB841;
    color: white;
    cursor: pointer;
    margin: 0px auto;
    transition: 0.3s;
    font-size: 20px;
    padding: 10px;
    &:hover {
      color: white;
      background-color: rgb(50, 50, 50);
    }
    &:disabled {
      background: rgb(152, 150, 150);
      color: black;
    }
  }
  
}