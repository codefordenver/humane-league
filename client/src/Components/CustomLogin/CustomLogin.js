import React, { Component } from 'react'
import './CustomLogin.css';
import firebase from '../../firebase';
import { connect } from 'react-redux';
import * as actions from '../../Actions/';
import { signInUser } from '../../utils/apiCalls';
import { Link } from 'react-router-dom';


class CustomLogin extends Component {
  constructor() {
    super();
    this.state = {
      email: '',
      password: '',
      confirmPass: '',
      name: '',
      error: null,
      signin: true,
      passMatchStatus: null,
      disableSubmit: true
    };
  }

  handleChange = (event) => {
    const { value, name } = event.target;
    this.setState({ [name]: value });

    if (this.state.signin) {
      if (this.email.value.length >= 5 && this.password.value.length >= 5) {
        this.setState({ disableSubmit: false });
      } else {
        this.setState({ disableSubmit: true });
      }
    }

    if (!this.state.signin) {
      if (this.password.value !== this.confirmPass.value || this.email.value.length < 5 || this.password.value.length < 5) {
        console.log('disabled');
        this.setState({ disableSubmit: true });
      } else {
        console.log('enabled');
        this.setState({ disableSubmit: false });
      }

      if (event.target.name === 'confirmPass' || event.target.name === 'password') {
        if (this.password.value === this.confirmPass.value && event.target.value.length)  {
          this.setState({ passMatchStatus: '✅' });
        } else {
          this.setState({ passMatchStatus: '❌' });
        }
      }
    }




  }

  signinHandler = () => {
    const { email, password } = this.state;
    firebase.auth().signInWithEmailAndPassword(email, password).catch( (error) => {
      if (error.code === "auth/user-not-found" || error.code === "auth/wrong-password") {
        this.setState({
          error: 'Invalid username or password. Please try again.'
        });

        setTimeout(() => {
          this.setState({ error: '' });
        }, 3000);
      }

      if (error.code === "auth/invalid-email") {
        this.setState({
          error: 'Invalid email.'
        });

        setTimeout(() => {
          this.setState({ error: '' });
        }, 3000);
      }
    });

    firebase.auth().onAuthStateChanged( async user => {
      console.log("auth change in signinhandler - usersigned in? ", !!user);
      if (user) {
        const { displayName, email, uid } = user;

        const dbUser = await signInUser(uid, displayName, email);
  
        this.props.validateUser(dbUser);
        localStorage.setItem('THL-FAN-USER', JSON.stringify(dbUser));
        this.props.history.push('/home');
      } else {
      }
    });
  }

  signupHandler = () => {
    const { email, password } = this.state;
    console.log('whatup');
    firebase.auth().createUserWithEmailAndPassword(email, password).catch(error => {
      if (error.code === "auth/invalid-email") {
        this.setState({
          error: 'Invalid email.'
        });
      } else {
        this.setState({
          error: 'Email already in use.'
        });
      }

      setTimeout(() => {
        this.setState({ error: '' });
      }, 3000);
    });

    firebase.auth().onAuthStateChanged(async user => {
      console.log("auth change in signuphandler - usersigned in? ", !!user);  
      if (user) {
        firebase.auth().currentUser.updateProfile({displayName: this.state.name});

        const { email, uid } = firebase.auth().currentUser;

        const dbUser = await signInUser(uid, this.state.name, email);
  
        this.props.validateUser(dbUser);
        localStorage.setItem('THL-FAN-USER', JSON.stringify(dbUser));
        this.props.history.push('/home');
      }
    });
  }

  createForm = () => {
    if (this.state.signin) {
      return (
        <div className="CustomLogin">
        <h1>Sign in with Email</h1>
        <label>Email:</label>
        <input 
          value={this.state.email}
          ref={(input) => { this.email = input; }} 
          name="email" 
          onChange={this.handleChange} 
          type="email" 
          placeholder="Email"/>
        <label>Password:</label>
        <input 
          value={this.state.password}
          ref={(input) => { this.password = input }} 
          name="password" 
          onChange={this.handleChange} 
          type="password" 
          placeholder="Password"/>
        <Link to="/forgotpassword">Forgot Password</Link>
        <span>{this.state.error}</span>
        <button disabled={this.state.disableSubmit} ref={(button) => { this.submit = button; }} onClick={this.signinHandler}>Sign In</button>
        <button onClick={() => {this.setState({signin: !this.state.signin})}}>Sign Up Instead</button>
        
      </div>
      )
    } else {
      return (
        <div className="CustomLogin">
        <h1>Sign Up with Email</h1>
        <label>Name:</label>
        <input 
          ref={(input) => { this.name = input; }} 
          name="name" 
          onChange={this.handleChange} 
          type="text" 
          value={this.state.name}
          placeholder="First/Last Name"/>
        <label>Email:</label>
        <input 
          ref={(input) => { this.email = input; }} 
          name="email" 
          onChange={this.handleChange} 
          type="email" 
          value={this.state.email}
          placeholder="Email"/>
        <label>Password:</label>
        <input 
          ref={(input) => { this.password = input }} 
          name="password" 
          onChange={this.handleChange} 
          value={this.state.password}
          type="password" 
          placeholder="Password"/>
        <label>Confirm Password:</label>
        <div className="confirmWrapper">
          <input 
            ref={(input) => { this.confirmPass = input }} 
            name="confirmPass" 
            onChange={this.handleChange} 
            value={this.state.confirmPass}
            type="password" 
            placeholder="Confirm Password" />
          <span>{this.state.passMatchStatus}</span>        
        </div>
        <button disabled={this.state.disableSubmit} ref={(button) => { this.submit = button; }} onClick={this.signupHandler}>Sign Up</button>
        <button onClick={() => {this.setState({signin: !this.state.signin})}}>Sign In Instead</button>
      </div>
      )
    }
  }

  render() {
    return this.createForm();
  }
}

const mapStateToProps = store => ({
  User: store.User
});

const mapDispatchToProps = dispatch => ({
  validateUser: user => dispatch(actions.updateUser(user))
});



export default connect(mapStateToProps, mapDispatchToProps)(CustomLogin);