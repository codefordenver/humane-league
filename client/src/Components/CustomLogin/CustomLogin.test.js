import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import { CustomLogin } from './CustomLogin';

describe('CustomLogin component tests', () => {
  let renderedCustomLogin;
  let defaultState = {
    email: '',
    password: '',
    confirmPass: '',
    name: '',
    error: null,
    signin: true,
    passMatchStatus: null,
    disableSubmit: true
  }
  
  beforeEach(() => {
    renderedCustomLogin = shallow( <CustomLogin /> );
  })
  it('renders without crashing', () => {
    expect(renderedCustomLogin).toBeDefined();
    expect(renderedCustomLogin.state()).toEqual(defaultState);
  });
  it('matches SnapShot', () => {
    expect(renderedCustomLogin).toMatchSnapshot();
  })
})