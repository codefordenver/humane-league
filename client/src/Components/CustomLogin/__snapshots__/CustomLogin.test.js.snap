// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomLogin component tests matches SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CustomLogin />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <h1>
          Sign in with Email
</h1>,
        <label>
          Email:
</label>,
        <input
          autoComplete="email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="email"
          value=""
/>,
        <label>
          Password:
</label>,
        <input
          autoComplete="current-password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
/>,
        <Link
          replace={false}
          to="/forgotpassword"
>
          Forgot Password
</Link>,
        <span />,
        <button
          disabled={true}
          onClick={[Function]}
>
          Sign In
</button>,
        <button
          onClick={[Function]}
>
          Sign Up Instead
</button>,
      ],
      "className": "CustomLogin",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign in with Email",
        },
        "ref": null,
        "rendered": "Sign in with Email",
        "type": "h1",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Email:",
        },
        "ref": null,
        "rendered": "Email:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoComplete": "email",
          "name": "email",
          "onChange": [Function],
          "placeholder": "Email",
          "type": "email",
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Password:",
        },
        "ref": null,
        "rendered": "Password:",
        "type": "label",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "autoComplete": "current-password",
          "name": "password",
          "onChange": [Function],
          "placeholder": "Password",
          "type": "password",
          "value": "",
        },
        "ref": [Function],
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": "Forgot Password",
          "replace": false,
          "to": "/forgotpassword",
        },
        "ref": null,
        "rendered": "Forgot Password",
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": null,
        },
        "ref": null,
        "rendered": null,
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign In",
          "disabled": true,
          "onClick": [Function],
        },
        "ref": [Function],
        "rendered": "Sign In",
        "type": "button",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign Up Instead",
          "onClick": [Function],
        },
        "ref": null,
        "rendered": "Sign Up Instead",
        "type": "button",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <h1>
            Sign in with Email
</h1>,
          <label>
            Email:
</label>,
          <input
            autoComplete="email"
            name="email"
            onChange={[Function]}
            placeholder="Email"
            type="email"
            value=""
/>,
          <label>
            Password:
</label>,
          <input
            autoComplete="current-password"
            name="password"
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
/>,
          <Link
            replace={false}
            to="/forgotpassword"
>
            Forgot Password
</Link>,
          <span />,
          <button
            disabled={true}
            onClick={[Function]}
>
            Sign In
</button>,
          <button
            onClick={[Function]}
>
            Sign Up Instead
</button>,
        ],
        "className": "CustomLogin",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign in with Email",
          },
          "ref": null,
          "rendered": "Sign in with Email",
          "type": "h1",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Email:",
          },
          "ref": null,
          "rendered": "Email:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoComplete": "email",
            "name": "email",
            "onChange": [Function],
            "placeholder": "Email",
            "type": "email",
            "value": "",
          },
          "ref": [Function],
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Password:",
          },
          "ref": null,
          "rendered": "Password:",
          "type": "label",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "autoComplete": "current-password",
            "name": "password",
            "onChange": [Function],
            "placeholder": "Password",
            "type": "password",
            "value": "",
          },
          "ref": [Function],
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": "Forgot Password",
            "replace": false,
            "to": "/forgotpassword",
          },
          "ref": null,
          "rendered": "Forgot Password",
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": null,
          },
          "ref": null,
          "rendered": null,
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign In",
            "disabled": true,
            "onClick": [Function],
          },
          "ref": [Function],
          "rendered": "Sign In",
          "type": "button",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign Up Instead",
            "onClick": [Function],
          },
          "ref": null,
          "rendered": "Sign Up Instead",
          "type": "button",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
